Adding CosmosDB managed resources.

artursouza opened the pull request:
Draft

artursouza committed:
Adding CosmosDB Account managed resource.

Signed-off-by: Artur Souza <artursouza.ms@outlook.com>

artursouza commented on apis/database/v1beta1/nosql_types.go:
```diff
@@ -0,0 +1,290 @@
+// +kubebuilder:object:root=true
+
+// A CosmosDBAccount is a managed resource that represents an Azure CosmosDB account with CosmosDB API.
+// +kubebuilder:printcolumn:name="STATUS",type="string",JSONPath=".status.bindingPhase"
```
status and state are not displayed. The jsonPath does not work. It is the same as used for MySQL and Postgres, so it seems to be a generalized problem in this provider's code. I will fix it in a separate PR.

hasheddan replied:
`bindingPhase` will likely not appear unless you actually claim the resource. Have you verified that it is still blank when you claim? `status.state` below is not a field in `runtimev1alpha1.ResourceSpec` so I would not expect it to be exposed. I'll take a look at `MySQL` and `Postgres` to see if they are broken :+1:

artursouza commented on apis/database/v1beta1/nosql_types.go:
```diff
@@ -0,0 +1,290 @@
+
+// A NoSQLAccountClassSpecTemplate is a template for the spec of a dynamically
+// provisioned CosmosDB or Cassandra via CosmosDB.
+type NoSQLAccountClassSpecTemplate struct {
```
Remove since composition will replace class and allow 1 to many mapping.

artursouza commented on apis/database/v1beta1/nosql_types.go:
```diff
@@ -0,0 +1,290 @@
+}
+
+// NoSQLAccountParameters define the desired state of an Azure CosmosDB account.
+type NoSQLAccountParameters struct {
```
CosmosDBAccountParams. CosmosDB prefix everywhere.

artursouza commented on apis/database/v1beta1/nosql_types.go:
```diff
@@ -0,0 +1,290 @@
+	// ConsistencyPolicy - The consistency policy for the Cosmos DB account.
+	ConsistencyPolicy *ConsistencyPolicy `json:"consistencyPolicy,omitempty"`
+	// Locations - An array that contains the georeplication locations enabled for the Cosmos DB account.
+	Locations *[]Location `json:"locations,omitempty"`
```
remove omitempty since this is required.

artursouza commented on apis/database/v1beta1/nosql_types.go:
```diff
@@ -0,0 +1,290 @@
+	ID *string `json:"id,omitempty"`
+
+	// Name - The name of the database account.
+	Name *string `json:"name,omitempty"`
```
remove omitempty on required fields.

artursouza commented on apis/database/v1beta1/nosql_types.go:
```diff
@@ -0,0 +1,290 @@
+
+	// Identity - The identity of the resource.
+	// +optional
+	ID *string `json:"id,omitempty"`
```
Keep pointers only for optional attributes.

artursouza commented on apis/database/v1beta1/register.go:
```diff
@@ -37,6 +37,22 @@ var (
+	CosmosDBAccountGroupVersionKind = SchemeGroupVersion.WithKind(CosmosDBAccountKind)
+)
+
+// NoSQLAccountClass type metadata.
```
Remove class.

artursouza commented on apis/database/v1beta1/register.go:
```diff
@@ -65,4 +81,7 @@ func init() {
 	SchemeBuilder.Register(&SQLServerClass{}, &SQLServerClassList{})
+
+	SchemeBuilder.Register(&CosmosDBAccount{}, &CosmosDBAccountList{})
+	SchemeBuilder.Register(&NoSQLAccountClass{}, &NoSQLAccountClassList{})
```
Remove class.

artursouza commented on config/crd/database.azure.crossplane.io_cosmosdbaccounts.yaml:
```diff
@@ -0,0 +1,411 @@
+  - JSONPath: .status.state
+    name: STATE
+    type: string
+  - JSONPath: .spec.classRef.name
```
Delete from CosmosDBAccount.

artursouza commented on pkg/clients/database/cosmosdb/account.go:
```diff
@@ -0,0 +1,147 @@
+	return &client, nil
+}
+
+/*
```
Remove commented out code.

artursouza commented on pkg/controller/database/cosmosdb/managed.go:
```diff
@@ -0,0 +1,150 @@
+	r.Status.SetConditions(runtimev1alpha1.Creating())
+	_, err := e.client.CreateOrUpdate(ctx,
+		r.Spec.ForProvider.ResourceGroupName,
+		*r.Spec.ForProvider.Name,
```
Use external name annotation. See AzureResourceGroup.

artursouza commented on pkg/controller/database/cosmosdb/managed.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+func (e *external) Update(_ context.Context, _ resource.Managed) (managed.ExternalUpdate, error) {
+	// TODO(negz): Support updates, if applicable.
```
Use my alias.

artursouza commented on pkg/controller/database/cosmosdb/managed.go:
```diff
@@ -0,0 +1,150 @@
+
+func (e *external) Update(_ context.Context, _ resource.Managed) (managed.ExternalUpdate, error) {
+	// TODO(negz): Support updates, if applicable.
+	return managed.ExternalUpdate{}, nil
```
Use CreatrOrUpdate API.

artursouza commented on pkg/controller/database/cosmosdb/managed.go:
```diff
@@ -0,0 +1,150 @@
+	}
+
+	r.SetConditions(runtimev1alpha1.Available())
+	return managed.ExternalObservation{ResourceExists: true, ResourceUpToDate: true}, nil
```
Set ResourceUpToDate to false if resource can safely be updated. Have a check function to confirm there was a change.

negz committed:
Bump runtime and tools

Signed-off-by: Nic Cope <negz@rk0n.org>

negz committed:
Remove reference resolution code from storage controllers

Azure storage types don't currently support references. It's likely we'll
migrate the controllers to managed.Reconciler before we add them.

Signed-off-by: Nic Cope <negz@rk0n.org>

negz committed:
Migrate to simple reference resolution and selection

This adds a reference selector for all existing references, and uses the new
managed.APISimpleReferenceResolver to select and resolve references.

Signed-off-by: Nic Cope <negz@rk0n.org>

negz committed:
Update Redis example to create a resource group

Signed-off-by: Nic Cope <negz@rk0n.org>

negz committed:
Merge pull request #137 from negz/referential

Migrate to simple reference resolution and selection
